### PR TITLE
osd: report numa information for network and storage; attempt to set affinity magically

### PR DESCRIPTION
**Pull request by liewegas**
- osds report numa node for network and storage
- new 'ceph osd numa-status' command
- new osd_numa_node option to make an osd bind (takes affect when OSD marks itself up)
- network IP selectoin prefers IPs in public_network that are on interfaces on the same numa node as the storage
- new osd_numa_auto_affinity option that sets affinity magically if network and storage are on the same numanode

An example from my dev box.  Unfortunately my NICs and NVMe are on different numa nodes, so I can't fully demo the auto-affinity behavior, but this does show explicitly setting the osd_numa_node setting.
```
gnit:build (wip-numa) 01:12 PM $ bin/ceph osd numa-status
OSD HOST NETWORK STORAGE AFFINITY CPUS 
  0 gnit 0       1       -        -    
  1 gnit 0       1       -        -    
  2 gnit 0       1       -        -    
gnit:build (wip-numa) 01:12 PM $ bin/ceph config set osd.0 osd_numa_node 0
gnit:build (wip-numa) 01:12 PM $ bin/ceph config set osd.1 osd_numa_node 1
gnit:build (wip-numa) 01:12 PM $ bin/ceph osd down 0 1
marked down osd.0. marked down osd.1. 
gnit:build (wip-numa) 01:12 PM $ bin/ceph osd numa-status
OSD HOST NETWORK STORAGE AFFINITY CPUS       
  0 gnit 0       1       0        0-7,16-23  
  1 gnit 0       1       1        8-15,24-31 
  2 gnit 0       1       -        -          
gnit:build (wip-numa) 01:16 PM $ taskset  -pc `cat out/osd.0.pid`
pid 31315's current affinity list: 0-7,16-23
gnit:build (wip-numa) 01:16 PM $ taskset  -pc `cat out/osd.1.pid`
pid 31644's current affinity list: 8-15,24-31
gnit:build (wip-numa) 01:16 PM $ taskset  -pc `cat out/osd.2.pid`
pid 31976's current affinity list: 0-31
```
also note that there is a bunch of numa information now in the osd metadata:
```
gnit:build (wip-numa) 01:14 PM $ bin/ceph osd metadata 0 | grep numa
    "network_numa_node": "0",
    "network_numa_nodes": "0",
    "numa_node": "0",
    "numa_node_cpus": "0-7,16-23",
    "objectstore_numa_node": "1",
    "objectstore_numa_nodes": "1",
```

### CODE REVIEW

**Comment by liewegas** [comment]
@ksingh7

**Comment by liewegas** [comment]
retest this please

**Comment by markhpc** [comment]
installing this on incerta to try it out!

File: src/common/numa.cc [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  }
+  return 'a' + (v - 10);
+}
+
+// FIXME: this is assuming the local_cpus value is little endian... is
+// that right?
+
+// FIXME: this assumes the cpu count is a multiple of 4 (a nibble)
+
+int parse_cpu_set_mask(
```
Comment (markhpc) [inline src/common/numa.cc]: Is this code actually used anywhere other than in the tests?

**Comment by liewegas** [comment]
nope.  i was originally parsing the local_cpus file but switched to local_cpulist instead.  could remove them

**Comment by markhpc** [comment]
Attempted to run this on Incerta last night.  On incerta we use the public/cluster network ceph.conf tunables to make sure we are using the 40GbE interface for the OSDs:
> public network = 10.0.10.0/24
> cluster network = 10.0.10.0/24

When attempting to run the wip-numa branch, the OSDs error out:
> 2019-01-08 07:44:05.432 7fb8ee6c7a40 -1 unable to find any IP address in networks '10.0.10.0/24' interfaces ''
> 2019-01-08 07:44:05.432 7fb8ee6c7a40 -1 Failed to pick public address.


while ifconfig clearly shows that one exists:
> ens513f1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 9000
>        inet 10.0.10.101  netmask 255.255.255.0  broadcast 10.0.10.255
>        inet6 fe80::6a05:caff:fe2b:7ea1  prefixlen 64  scopeid 0x20<link>
>        ether 68:05:ca:2b:7e:a1  txqueuelen 1000  (Ethernet)
>        RX packets 3677983381  bytes 17667220742426 (16.0 TiB)
>        RX errors 0  dropped 0  overruns 0  frame 0
>        TX packets 4407351006  bytes 30510656493544 (27.7 TiB)
>        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

**Comment by markhpc** [comment]
Good job, works on incerta for me now as well:

```
OSD HOST      NETWORK STORAGE AFFINITY CPUS 
  0 incerta01       0       1        - -    
  1 incerta01       0       1        - -    
  2 incerta01       0       1        - -    
  3 incerta01       0       1        - -    
  4 incerta02       0       1        - -    
  5 incerta02       0       1        - -    
  6 incerta02       0       1        - -    
  7 incerta02       0       1        - -    
  8 incerta03       0       1        - -    
  9 incerta03       0       1        - -    
 10 incerta03       0       1        - -    
 11 incerta03       0       1        - -    
 12 incerta04       0       1        - -    
 13 incerta04       0       1        - -    
 14 incerta04       0       1        - -    
 15 incerta04       0       1        - -    

```

Looking at this, what would you think about some kind of greedy aproach?  IE if we don't know what to do, pin to the least used numa node?

**Comment by liewegas** [comment]
@markhpc I don't think we should try to do any pinning automatically unless we are very certain it is a good idea.  I suggest including this for nautilus, which gives users an easy way to see and explicitly set numa pinning, and see what we learn from it.

**Comment by liewegas** [comment]
i.e., "we don't know what to do" should always mean do nothing.

**Comment by markhpc** [comment]
That's sort of the point of my question though.  Say half the drives in the node use a topology that we can easily make a decision on (network and disk are on the same numa node).  We pin those.  The other half have their disk and network split.  Are we better off leaving those nodes unpinned and just hoping that the kernel puts them on the other node and doesn't try to move them around or are we better off pinning them to the other node since the first is already full of OSDs?

**Comment by liewegas** [comment]
IMO the fact that the answer isn't obvious (and probably changes between a hyperconverges vs dedicated storage node) means we shouldn't, at least not at this stage.

**Pull request merged**